noting ข้างล่างจะเป็นการเปิดไฟล์และตัดตารางไฟล์เพื่อเอาข้อมูลสำคัญออกมา

In [ ]:
import pandas as pd

### Loading `/content/NPASS3.0_naturalproducts_generalinfo.txt`

natural prodruct เอาข้อมูลแค่ np id ,pref name ,iupac name

In [ ]:
df_generalinfo = pd.read_csv('/content/NPASS3.0_naturalproducts_generalinfo.txt', sep='\t')
display(df_generalinfo.head())

,np_id,inchikey,pref_name,iupac_name,chembl_id,pubchem_id,name_initial,num_of_organism,num_of_target,num_of_activity,gene_cluster,ifQuantity
0,NPC491451,NLKSRZQAJMOBTP-UHFFFAOYSA-N,(Z)-1-hydroxynon-7-en-5-yn-4-one,1-hydroxynon-7-en-5-yn-4-one,n.a.,n.a.,(,1.0,0.0,0.0,n.a.,No
1,NPC491452,OZLLMLZOMPDONB-UHFFFAOYSA-N,4-(2-methoxy-5-methoxycarbonyl-phenoxy)benzoic...,4-(2-methoxy-5-methoxycarbonyl-phenoxy)benzoic...,n.a.,n.a.,4,1.0,0.0,0.0,n.a.,No
2,NPC491453,FELYAZAWTURXNF-ODUVEOQFSA-N,Nanchangmycin,"(~{E},2~{S},4~{R},8~{S})-8-[(2~{S},5~{R},7~{S}...",n.a.,n.a.,N,4.0,0.0,0.0,n.a.,No
3,NPC491454,QSTDRAXWKGCQCD-UHFFFAOYSA-N,"2-(2-hydroxy-5-methoxy-phenyl)-1,3-benzoxazole...","2-(2-hydroxy-5-methoxy-phenyl)-1,3-benzoxazole...",n.a.,n.a.,2,1.0,0.0,0.0,n.a.,No
4,NPC491455,XOAQRSFHWAISAM-GOSBJONWSA-N,phomopsichalasin D,"(1~{S},2~{R},5~{R},7~{S},11~{S},14~{S},15~{R},...",n.a.,n.a.,p,1.0,0.0,0.0,n.a.,No


In [ ]:
df_natural_products_selected = df_generalinfo[['np_id', 'pref_name', 'iupac_name']]

display(df_natural_products_selected.head())

df_natural_products_selected.to_csv('natural_products_selected.csv', index=False)
print('File natural_products_selected.csv has been saved.')

,np_id,pref_name,iupac_name
0,NPC491451,(Z)-1-hydroxynon-7-en-5-yn-4-one,1-hydroxynon-7-en-5-yn-4-one
1,NPC491452,4-(2-methoxy-5-methoxycarbonyl-phenoxy)benzoic...,4-(2-methoxy-5-methoxycarbonyl-phenoxy)benzoic...
2,NPC491453,Nanchangmycin,"(~{E},2~{S},4~{R},8~{S})-8-[(2~{S},5~{R},7~{S}..."
3,NPC491454,"2-(2-hydroxy-5-methoxy-phenyl)-1,3-benzoxazole...","2-(2-hydroxy-5-methoxy-phenyl)-1,3-benzoxazole..."
4,NPC491455,phomopsichalasin D,"(1~{S},2~{R},5~{R},7~{S},11~{S},14~{S},15~{R},..."


File natural_products_selected.csv has been saved.


### Loading `/content/NPASS3.0_naturalproducts_species_pair.txt`

naturalproduct species pair เอา org ID,np id

In [ ]:
df_species_pair = pd.read_csv('/content/NPASS3.0_naturalproducts_species_pair.txt', sep='\t')
display(df_species_pair.head())

,src_org_record_id,src_org_pair_id,src_org_pair,org_id,np_id,new_cp_found,org_isolation_part,org_collect_location,org_collect_time,ref_type,ref_id,ref_id_type,ref_url
0,1417289,681673,NPO27288-NPC251357,NPO27288,NPC251357,n.a.,n.a.,n.a.,n.a.,Database,TM-MC,Database,n.a.
1,1538850,774177,NPO8589-NPC116013,NPO8589,NPC116013,N,n.a.,"Tsukuba, Japan",1998-JUN,Publication,12381108,PMID,https://pubmed.ncbi.nlm.nih.gov/12381108
2,1084309,352357,NPO14625-NPC39,NPO14625,NPC39,n.a.,n.a.,n.a.,n.a.,Database,UNPD,Database,n.a.
3,946868,215202,NPO16590-NPC190457,NPO16590,NPC190457,n.a.,n.a.,n.a.,n.a.,Database,UNPD,Database,n.a.
4,1137261,405190,NPO6861-NPC13143,NPO6861,NPC13143,n.a.,n.a.,n.a.,n.a.,Database,UNPD,Database,n.a.


In [ ]:
df_species_pair_selected = df_species_pair[['org_id', 'np_id']]
display(df_species_pair_selected.head())

,org_id,np_id
0,NPO27288,NPC251357
1,NPO8589,NPC116013
2,NPO14625,NPC39
3,NPO16590,NPC190457
4,NPO6861,NPC13143


In [ ]:
df_species_pair_selected.to_csv('species_pair_selected.csv', index=False)
print('File species_pair_selected.csv has been saved.')

File species_pair_selected.csv has been saved.


### Loading `/content/NPASS3.0_naturalproducts_structure.txt`

อาเชื่อมฐานเคมี chembl เพื่อเอาไอดีที่แมชกันแล้วเก็บไว้ในตาราง และตารางสุดท้ายเอาเพียงแค่ np id , standard inchi ,standard inchi key,chembl id

In [ ]:
df_structure = pd.read_csv('/content/NPASS3.0_naturalproducts_structure.txt', sep='\t')
display(df_structure.head())

,np_id,InChI,InChIKey,SMILES
0,NPC81384,"InChI=1S/C8H20NO6P/c1-9(2,3)4-5-14-16(12,13)15...",SUHOQUVVVLNYQR-MRVPVSSYSA-N,OC[C@H](COP(=O)(OCC[N+](C)(C)C)[O-])O
1,NPC109991,InChI=1S/C24H21NO5/c1-3-13(2)21(24(29)30)25-12...,RVOHRYJJXCFNQS-YEJXKQKISA-N,CC[C@H](C)[C@H](C(=O)O)n1cc2C=C(C(=O)c3c(ccc(c...
2,NPC4649,InChI=1S/C20H20O7/c1-5-6-14(22)26-19-18(24-11(...,MHLKYRDXNNGAMX-WLMAVBAVSA-N,C/C=C/C(=O)O[C@H]1[C@H](c2c(ccc3ccc(=O)oc23)OC...
3,NPC90844,InChI=1S/C23H25NO3/c1-25-15-6-7-16-18(10-15)20...,NCVWJDISIZHFQS-AWEZNQCLSA-N,COc1ccc2c(c1)c1cc(c(cc1c1C[C@@H]3CCCN3Cc21)OC)OC
4,NPC18064,InChI=1S/C30H48O3/c1-18-10-15-30(25(32)33)17-1...,HNJKFVJZKCXLRL-SILQXMKOSA-N,CC1=CC[C@]2([C@H]([C@@H]1C)[C@H]1CC[C@H]3[C@@]...


In [ ]:
!pip install chembl_downloader
import pandas as pd
import sqlite3

# 1. Load NPASS Natural Products Structure Data
try:
    # Assuming the file is now consistently in /content/sample_data/
    npass_structure_df = pd.read_csv(
        "/content/sample_data/NPASS3.0_naturalproducts_structure.txt",
        sep="\t",
        header=0
    )
    print("NPASS Natural Products Structure data loaded successfully.")
    print(f"Shape: {npass_structure_df.shape}")
    print("First 5 rows of NPASS data:")
    display(npass_structure_df.head())
except FileNotFoundError:
    print("Error: NPASS3.0_naturalproducts_structure.txt not found. Please ensure it's in the correct path.")
    npass_structure_df = pd.DataFrame()

# Ensure chembl_path is defined from previous steps
if 'chembl_path' not in locals():
    print("Warning: 'chembl_path' not found. Please run the ChEMBL downloader cell first.")
    # Fallback if chembl_path is not defined (e.g. if the previous cell was skipped)
    import chembl_downloader
    chembl_path = chembl_downloader.download_extract_sqlite(version="36")
    print(f"ChEMBL database path: {chembl_path}")

# 2. Query ChEMBL Database
if chembl_path:
    conn = sqlite3.connect(chembl_path)
    chembl_query = """
    SELECT
        md.chembl_id,
        cs.standard_inchi,
        cs.standard_inchi_key
    FROM
        molecule_dictionary md
    JOIN
        compound_structures cs ON md.molregno = cs.molregno
    """
    chembl_df = pd.read_sql_query(chembl_query, conn)
    conn.close()
    print("\nChEMBL data queried successfully.")
    print(f"Shape: {chembl_df.shape}")
    print("First 5 rows of ChEMBL data:")
    display(chembl_df.head())
else:
    print("Cannot query ChEMBL: chembl_path is not defined or invalid.")
    chembl_df = pd.DataFrame()

# 3. Merge DataFrames
if not npass_structure_df.empty and not chembl_df.empty:
    # Ensure column types are consistent for merging (e.g., stripping whitespace)
    npass_structure_df['InChI'] = npass_structure_df['InChI'].astype(str).str.strip()
    npass_structure_df['InChIKey'] = npass_structure_df['InChIKey'].astype(str).str.strip()
    chembl_df['standard_inchi'] = chembl_df['standard_inchi'].astype(str).str.strip()
    chembl_df['standard_inchi_key'] = chembl_df['standard_inchi_key'].astype(str).str.strip()

    # Merge on InChI and InChIKey (using standard_inchi and standard_inchi_key from ChEMBL for matching)
    merged_df = pd.merge(
        npass_structure_df,
        chembl_df,
        left_on=['InChI', 'InChIKey'],
        right_on=['standard_inchi', 'standard_inchi_key'],
        how='left' # Use left merge to keep all NPASS compounds and add ChEMBL info where available
    )

    # 4. Select Final Columns
    # The user asked for 'inchi' and 'inchikey' which are present in both, using NPASS's original here
    final_df = merged_df[[
        'np_id',
        'InChI', # NPASS original InChI
        'InChIKey', # NPASS original InChIKey
        'chembl_id',
        'standard_inchi',
        'standard_inchi_key'
    ]].copy()

    print("\nMerged DataFrame created successfully.")
    print(f"Shape: {final_df.shape}")
    print("First 5 rows of the new DataFrame with ChEMBL info:")
    display(final_df.head())
    print("\nValue counts for chembl_id (top 10):\n", final_df['chembl_id'].value_counts().head(10))
    print("\nNumber of NPASS compounds successfully mapped to ChEMBL IDs:", final_df['chembl_id'].nunique())
else:
    print("Merged DataFrame could not be created due to empty input DataFrames.")
    final_df = pd.DataFrame()

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.4/46.4 kB 3.4 MB/s eta 0:00:00
Error: NPASS3.0_naturalproducts_structure.txt not found. Please ensure it's in the correct path.


ChEMBL database path: /root/.data/chembl/36/chembl_36.db

ChEMBL data queried successfully.
Shape: (2854815, 3)
First 5 rows of ChEMBL data:


,chembl_id,standard_inchi,standard_inchi_key
0,CHEMBL6329,InChI=1S/C17H12ClN3O3/c1-10-8-11(21-17(24)20-1...,OWRSAHYFSSNENM-UHFFFAOYSA-N
1,CHEMBL6328,InChI=1S/C18H12N4O3/c1-11-8-14(22-18(25)21-16(...,ZJYUMURGSZQFMH-UHFFFAOYSA-N
2,CHEMBL265667,InChI=1S/C18H16ClN3O3/c1-10-7-14(22-18(25)21-1...,YOMWDCALSDWFSV-UHFFFAOYSA-N
3,CHEMBL6362,InChI=1S/C17H13N3O3/c1-11-2-4-12(5-3-11)16(22)...,PSOPUAQFGCRDIP-UHFFFAOYSA-N
4,CHEMBL267864,InChI=1S/C17H12ClN3O3/c1-10-8-13(21-17(24)20-1...,KEZNSCMBVRNOHO-UHFFFAOYSA-N


Merged DataFrame could not be created due to empty input DataFrames.


In [ ]:
import pandas as pd
import sqlite3

# 1. Load NPASS Natural Products Structure Data
try:
    # Corrected file path
    npass_structure_df = pd.read_csv(
        "/content/NPASS3.0_naturalproducts_structure.txt",
        sep="\t",
        header=0
    )
    print("NPASS Natural Products Structure data loaded successfully.")
    print(f"Shape: {npass_structure_df.shape}")
    print("First 5 rows of NPASS data:")
    display(npass_structure_df.head())
except FileNotFoundError:
    print("Error: NPASS3.0_naturalproducts_structure.txt not found. Please ensure it's in the correct path.")
    npass_structure_df = pd.DataFrame()

# Ensure chembl_path is defined from previous steps
if 'chembl_path' not in locals():
    print("Warning: 'chembl_path' not found. Please run the ChEMBL downloader cell first.")
    # Fallback if chembl_path is not defined (e.g. if the previous cell was skipped)
    import chembl_downloader
    chembl_path = chembl_downloader.download_extract_sqlite(version="36")
    print(f"ChEMBL database path: {chembl_path}")

# 2. Query ChEMBL Database
if chembl_path:
    conn = sqlite3.connect(chembl_path)
    chembl_query = """
    SELECT
        md.chembl_id,
        cs.standard_inchi,
        cs.standard_inchi_key
    FROM
        molecule_dictionary md
    JOIN
        compound_structures cs ON md.molregno = cs.molregno
    """
    chembl_df = pd.read_sql_query(chembl_query, conn)
    conn.close()
    print("\nChEMBL data queried successfully.")
    print(f"Shape: {chembl_df.shape}")
    print("First 5 rows of ChEMBL data:")
    display(chembl_df.head())
else:
    print("Cannot query ChEMBL: chembl_path is not defined or invalid.")
    chembl_df = pd.DataFrame()

# 3. Merge DataFrames
if not npass_structure_df.empty and not chembl_df.empty:
    # Ensure column types are consistent for merging (e.g., stripping whitespace)
    npass_structure_df['InChI'] = npass_structure_df['InChI'].astype(str).str.strip()
    npass_structure_df['InChIKey'] = npass_structure_df['InChIKey'].astype(str).str.strip()
    chembl_df['standard_inchi'] = chembl_df['standard_inchi'].astype(str).str.strip()
    chembl_df['standard_inchi_key'] = chembl_df['standard_inchi_key'].astype(str).str.strip()

    # Merge on InChI and InChIKey (using standard_inchi and standard_inchi_key from ChEMBL for matching)
    merged_df = pd.merge(
        npass_structure_df,
        chembl_df,
        left_on=['InChI', 'InChIKey'],
        right_on=['standard_inchi', 'standard_inchi_key'],
        how='left' # Use left merge to keep all NPASS compounds and add ChEMBL info where available
    )

    # 4. Select Final Columns
    # The user asked for 'inchi' and 'inchikey' which are present in both, using NPASS's original here
    final_df = merged_df[[
        'np_id',
        'InChI', # NPASS original InChI
        'InChIKey', # NPASS original InChIKey
        'chembl_id',
        'standard_inchi',
        'standard_inchi_key'
    ]].copy()

    print("\nMerged DataFrame created successfully.")
    print(f"Shape: {final_df.shape}")
    print("First 5 rows of the new DataFrame with ChEMBL info:")
    display(final_df.head())
    print("\nValue counts for chembl_id (top 10):\n", final_df['chembl_id'].value_counts().head(10))
    print("\nNumber of NPASS compounds successfully mapped to ChEMBL IDs:", final_df['chembl_id'].nunique())
else:
    print("Merged DataFrame could not be created due to empty input DataFrames.")
    final_df = pd.DataFrame()

NPASS Natural Products Structure data loaded successfully.
Shape: (203390, 4)
First 5 rows of NPASS data:


,np_id,InChI,InChIKey,SMILES
0,NPC81384,"InChI=1S/C8H20NO6P/c1-9(2,3)4-5-14-16(12,13)15...",SUHOQUVVVLNYQR-MRVPVSSYSA-N,OC[C@H](COP(=O)(OCC[N+](C)(C)C)[O-])O
1,NPC109991,InChI=1S/C24H21NO5/c1-3-13(2)21(24(29)30)25-12...,RVOHRYJJXCFNQS-YEJXKQKISA-N,CC[C@H](C)[C@H](C(=O)O)n1cc2C=C(C(=O)c3c(ccc(c...
2,NPC4649,InChI=1S/C20H20O7/c1-5-6-14(22)26-19-18(24-11(...,MHLKYRDXNNGAMX-WLMAVBAVSA-N,C/C=C/C(=O)O[C@H]1[C@H](c2c(ccc3ccc(=O)oc23)OC...
3,NPC90844,InChI=1S/C23H25NO3/c1-25-15-6-7-16-18(10-15)20...,NCVWJDISIZHFQS-AWEZNQCLSA-N,COc1ccc2c(c1)c1cc(c(cc1c1C[C@@H]3CCCN3Cc21)OC)OC
4,NPC18064,InChI=1S/C30H48O3/c1-18-10-15-30(25(32)33)17-1...,HNJKFVJZKCXLRL-SILQXMKOSA-N,CC1=CC[C@]2([C@H]([C@@H]1C)[C@H]1CC[C@H]3[C@@]...



ChEMBL data queried successfully.
Shape: (2854815, 3)
First 5 rows of ChEMBL data:


,chembl_id,standard_inchi,standard_inchi_key
0,CHEMBL6329,InChI=1S/C17H12ClN3O3/c1-10-8-11(21-17(24)20-1...,OWRSAHYFSSNENM-UHFFFAOYSA-N
1,CHEMBL6328,InChI=1S/C18H12N4O3/c1-11-8-14(22-18(25)21-16(...,ZJYUMURGSZQFMH-UHFFFAOYSA-N
2,CHEMBL265667,InChI=1S/C18H16ClN3O3/c1-10-7-14(22-18(25)21-1...,YOMWDCALSDWFSV-UHFFFAOYSA-N
3,CHEMBL6362,InChI=1S/C17H13N3O3/c1-11-2-4-12(5-3-11)16(22)...,PSOPUAQFGCRDIP-UHFFFAOYSA-N
4,CHEMBL267864,InChI=1S/C17H12ClN3O3/c1-10-8-13(21-17(24)20-1...,KEZNSCMBVRNOHO-UHFFFAOYSA-N



Merged DataFrame created successfully.
Shape: (203390, 6)
First 5 rows of the new DataFrame with ChEMBL info:


,np_id,InChI,InChIKey,chembl_id,standard_inchi,standard_inchi_key
0,NPC81384,"InChI=1S/C8H20NO6P/c1-9(2,3)4-5-14-16(12,13)15...",SUHOQUVVVLNYQR-MRVPVSSYSA-N,CHEMBL1567463,"InChI=1S/C8H20NO6P/c1-9(2,3)4-5-14-16(12,13)15...",SUHOQUVVVLNYQR-MRVPVSSYSA-N
1,NPC109991,InChI=1S/C24H21NO5/c1-3-13(2)21(24(29)30)25-12...,RVOHRYJJXCFNQS-YEJXKQKISA-N,NaN,NaN,NaN
2,NPC4649,InChI=1S/C20H20O7/c1-5-6-14(22)26-19-18(24-11(...,MHLKYRDXNNGAMX-WLMAVBAVSA-N,NaN,NaN,NaN
3,NPC90844,InChI=1S/C23H25NO3/c1-25-15-6-7-16-18(10-15)20...,NCVWJDISIZHFQS-AWEZNQCLSA-N,CHEMBL483423,InChI=1S/C23H25NO3/c1-25-15-6-7-16-18(10-15)20...,NCVWJDISIZHFQS-AWEZNQCLSA-N
4,NPC18064,InChI=1S/C30H48O3/c1-18-10-15-30(25(32)33)17-1...,HNJKFVJZKCXLRL-SILQXMKOSA-N,CHEMBL517086,InChI=1S/C30H48O3/c1-18-10-15-30(25(32)33)17-1...,HNJKFVJZKCXLRL-SILQXMKOSA-N



Value counts for chembl_id (top 10):
 chembl_id
CHEMBL469036     3
CHEMBL3086842    3
CHEMBL457583     3
CHEMBL2361427    3
CHEMBL443231     3
CHEMBL3086841    3
CHEMBL461698     2
CHEMBL4533550    2
CHEMBL2359248    2
CHEMBL4636684    2
Name: count, dtype: int64

Number of NPASS compounds successfully mapped to ChEMBL IDs: 53376


In [ ]:
final_df.to_csv('natural_products_chembl_merged.csv', index=False)
print('File natural_products_chembl_merged.csv has been saved.')

File natural_products_chembl_merged.csv has been saved.


### Loading `/content/NPASS3.0_species_info.txt`

species info เอาแค่ org_id, org_name , org_tax_id,species_tax_id species_name genus_tax_id, genus_name,genus tax id,family tax id ,fam tax name,kingdom name,kingdom id,super kingdom ,superkingdom id

In [ ]:
df_species_info = pd.read_csv('/content/NPASS3.0_species_info.txt', sep='\t')
display(df_species_info.head())

,org_id,org_name,org_tax_level,org_tax_id,subspecies_tax_id,subspecies_name,species_tax_id,species_name,genus_tax_id,genus_name,...,kingdom_name,superkingdom_tax_id,superkingdom_name,org_name_initial,num_of_np_act,num_of_np_no_act,num_of_np_quantity,if_org_coculture,if_org_engineered,if_org_symbiont
0,NPO1,Astragalus flexuosus,Species,1234496,n.a.,n.a.,1234496,Astragalus flexuosus,20400,Astragalus,...,Viridiplantae,2759,Eukaryota,A,2,8,1,0,0,0
1,NPO10,Alexandrium ostenfeldii,Species,40986,n.a.,n.a.,40986,Alexandrium ostenfeldii,2924,Alexandrium,...,n.a.,2759,Eukaryota,A,10,43,0,0,0,0
2,NPO100,Zanthoxylum microcarpum,Species,n.a.,n.a.,n.a.,n.a.,n.a.,67937,Zanthoxylum,...,Viridiplantae,2759,Eukaryota,Z,1,9,1,0,0,0
3,NPO1000,Thymus transcaucasicus,Species,n.a.,n.a.,n.a.,n.a.,n.a.,49990,Thymus,...,Viridiplantae,2759,Eukaryota,T,19,29,6,0,0,0
4,NPO10001,Gnidia subcordata,Species,n.a.,n.a.,n.a.,n.a.,n.a.,142743,Gnidia,...,Viridiplantae,2759,Eukaryota,G,6,29,8,0,0,0


In [ ]:
df_species_info_selected = df_species_info[[
    'org_id', 'org_name', 'org_tax_id', 'species_tax_id', 'species_name',
    'genus_tax_id', 'genus_name', 'family_tax_id', 'family_name',
    'kingdom_name', 'kingdom_tax_id', 'superkingdom_name', 'superkingdom_tax_id'
]]
display(df_species_info_selected.head())

,org_id,org_name,org_tax_id,species_tax_id,species_name,genus_tax_id,genus_name,family_tax_id,family_name,kingdom_name,kingdom_tax_id,superkingdom_name,superkingdom_tax_id
0,NPO1,Astragalus flexuosus,1234496,1234496,Astragalus flexuosus,20400,Astragalus,3803,Fabaceae,Viridiplantae,33090,Eukaryota,2759
1,NPO10,Alexandrium ostenfeldii,40986,40986,Alexandrium ostenfeldii,2924,Alexandrium,1573500,Pyrocystaceae,n.a.,n.a.,Eukaryota,2759
2,NPO100,Zanthoxylum microcarpum,n.a.,n.a.,n.a.,67937,Zanthoxylum,23513,Rutaceae,Viridiplantae,33090,Eukaryota,2759
3,NPO1000,Thymus transcaucasicus,n.a.,n.a.,n.a.,49990,Thymus,4136,Lamiaceae,Viridiplantae,33090,Eukaryota,2759
4,NPO10001,Gnidia subcordata,n.a.,n.a.,n.a.,142743,Gnidia,39987,Thymelaeaceae,Viridiplantae,33090,Eukaryota,2759


In [ ]:
df_species_info_selected.to_csv('species_info_selected.csv', index=False)
print('File species_info_selected.csv has been saved.')

File species_info_selected.csv has been saved.


### Loading `/content/NPASS3.0_target.txt`

target เอาเชื่อมฐานข้อมูล chembl แล้วเอาแค่ target id, targetname, target type ,chembl

In [ ]:
df_target = pd.read_csv('/content/NPASS3.0_target.txt', sep='\t')
display(df_target.head())

,target_id,target_type,target_name,target_organism_tax_id,target_organism,uniprot_id
0,NPT1031,Cell line,Ca9-22,9606,Homo sapiens,n.a.
1,NPT1034,Cell line,Lu1,9606,Homo sapiens,n.a.
2,NPT1045,Cell line,U2OS,9606,Homo sapiens,n.a.
3,NPT1059,Cell line,HMEC,9606,Homo sapiens,n.a.
4,NPT1076,Cell line,B-cells,n.a.,n.a.,n.a.


In [ ]:
df_target_chembl_info = df_target_merged_chembl[[
    'target_id',
    'target_name',
    'target_type',
    'chembl_id'
]].copy()

print("Shape of df_target_chembl_info:", df_target_chembl_info.shape)
display(df_target_chembl_info.head())

Shape of df_target_chembl_info: (12700, 4)


,target_id,target_name,target_type,chembl_id
0,NPT1031,Ca9-22,Cell line,NaN
1,NPT1034,Lu1,Cell line,NaN
2,NPT1045,U2OS,Cell line,NaN
3,NPT1059,HMEC,Cell line,NaN
4,NPT1076,B-cells,Cell line,NaN


In [ ]:
df_target_chembl_info.to_csv('target_chembl_info.csv', index=False)
print('File target_chembl_info.csv has been saved.')

File target_chembl_info.csv has been saved.


file activities เอา compond ชน target

In [ ]:
import pandas as pd

df_activities = pd.read_csv('/content/sample_data/NPASS3.0_activities.txt', sep='\t')
display(df_activities.head())

/tmp/ipython-input-142514263.py:3: DtypeWarning: Columns (5) have mixed types. Specify dtype option on import or set low_memory=False.
  df_activities = pd.read_csv('/content/sample_data/NPASS3.0_activities.txt', sep='\t')


,np_id,target_id,activity_type_grouped,activity_relation,activity_type,activity_value,activity_units,assay_organism,assay_tax_id,assay_strain,assay_tissue,assay_cell_type,ref_id,ref_id_type
0,NPC46644,NPT918,Kd,=,Kd,26.0,nM,Homo sapiens,9606,n.a.,n.a.,n.a.,23362862,PMID
1,NPC229354,NPT918,Kd,=,Kd,16.0,nM,Homo sapiens,9606,n.a.,n.a.,n.a.,23362862,PMID
2,NPC477425,NPT211,IC50,=,IC50,820.0,nM,Homo sapiens,9606,n.a.,n.a.,T47D,18004798,PMID
3,NPC477427,NPT211,IC50,>,IC50,30000.0,nM,Homo sapiens,9606,n.a.,n.a.,T47D,18004798,PMID
4,NPC477531,NPT589,Others,=,Activity,47.0,%,Bos taurus,9913,n.a.,n.a.,n.a.,23691929,PMID


In [ ]:
df_activities_selected = df_activities[['np_id', 'target_id']]
display(df_activities_selected.head())

df_activities_selected.to_csv('activities_np_target_selected.csv', index=False)
print('File activities_np_target_selected.csv has been saved.')

,np_id,target_id
0,NPC46644,NPT918
1,NPC229354,NPT918
2,NPC477425,NPT211
3,NPC477427,NPT211
4,NPC477531,NPT589


File activities_np_target_selected.csv has been saved.


# Task
Merge the `df_structure` and `df_generalinfo` DataFrames on the `np_id` column using an inner merge to ensure only matching natural products are included. From the resulting DataFrame, select the `np_id`, `InChI`, `InChIKey`, and `chembl_id` columns. Rename the `InChI` column to `standard_inchi` and the `InChIKey` column to `standard_inchi_key`. Finally, display the first 5 rows of this new DataFrame and save it as a CSV file named `natural_products_chembl_info.csv` without including the index.

## Merge DataFrames for ChEMBL IDs

### Subtask:
Merge `df_structure` and `df_generalinfo` on `np_id` to combine structural information with ChEMBL IDs. This will create a comprehensive table containing all the necessary information for the final output.


**Reasoning**:
To combine structural information with general information including ChEMBL IDs, I will merge `df_structure` and `df_generalinfo` using the `np_id` column as the key.



In [ ]:
df_merged_chembl = pd.merge(df_structure, df_generalinfo, on='np_id', how='inner')
display(df_merged_chembl.head())

,np_id,InChI,InChIKey,SMILES,inchikey,pref_name,iupac_name,chembl_id,pubchem_id,name_initial,num_of_organism,num_of_target,num_of_activity,gene_cluster,ifQuantity
0,NPC81384,"InChI=1S/C8H20NO6P/c1-9(2,3)4-5-14-16(12,13)15...",SUHOQUVVVLNYQR-MRVPVSSYSA-N,OC[C@H](COP(=O)(OCC[N+](C)(C)C)[O-])O,SUHOQUVVVLNYQR-MRVPVSSYSA-N,Choline Alfoscerate,"[(2R)-2,3-dihydroxypropyl] 2-(trimethylazanium...",CHEMBL1567463,657272,C,9.0,67.0,86.0,n.a.,No
1,NPC109991,InChI=1S/C24H21NO5/c1-3-13(2)21(24(29)30)25-12...,RVOHRYJJXCFNQS-YEJXKQKISA-N,CC[C@H](C)[C@H](C(=O)O)n1cc2C=C(C(=O)c3c(ccc(c...,RVOHRYJJXCFNQS-YEJXKQKISA-N,RVOHRYJJXCFNQS-YEJXKQKISA-N,n.a.,n.a.,n.a.,R,6.0,0.0,0.0,n.a.,No
2,NPC4649,InChI=1S/C20H20O7/c1-5-6-14(22)26-19-18(24-11(...,MHLKYRDXNNGAMX-WLMAVBAVSA-N,C/C=C/C(=O)O[C@H]1[C@H](c2c(ccc3ccc(=O)oc23)OC...,MHLKYRDXNNGAMX-WLMAVBAVSA-N,MHLKYRDXNNGAMX-WLMAVBAVSA-N,n.a.,n.a.,5320691,M,1.0,0.0,0.0,n.a.,No
3,NPC90844,InChI=1S/C23H25NO3/c1-25-15-6-7-16-18(10-15)20...,NCVWJDISIZHFQS-AWEZNQCLSA-N,COc1ccc2c(c1)c1cc(c(cc1c1C[C@@H]3CCCN3Cc21)OC)OC,NCVWJDISIZHFQS-AWEZNQCLSA-N,Hypoestestatin 1,"(13aS)-2,3,6-trimethoxy-9,11,12,13,13a,14-hexa...",CHEMBL483423,12019847,H,1.0,4.0,19.0,n.a.,No
4,NPC18064,InChI=1S/C30H48O3/c1-18-10-15-30(25(32)33)17-1...,HNJKFVJZKCXLRL-SILQXMKOSA-N,CC1=CC[C@]2([C@H]([C@@H]1C)[C@H]1CC[C@H]3[C@@]...,HNJKFVJZKCXLRL-SILQXMKOSA-N,"3Beta-Hydroxy-18Alpha,19Alpha-Urs-20-En-28-Oic...","(1S,4aR,6aR,6aR,6bR,8aR,10S,12aR,14aR,14bR)-10...",CHEMBL517086,44575399,3,2.0,7.0,8.0,n.a.,No


**Reasoning**:
Now that the DataFrames are merged, I need to select the specified columns, rename them, display the head, and then save the resulting DataFrame to a CSV file as requested by the main task.



## Final Task

### Subtask:
Confirm that the task has been completed and the final DataFrame is ready for further use or inspection.


## Summary:

### Data Analysis Key Findings

*   `df_structure` and `df_generalinfo` DataFrames were successfully merged using an inner join on the `np_id` column, creating a combined DataFrame named `df_merged_chembl`.
*   A new DataFrame, `df_natural_products_chembl_info`, was created by selecting specific columns (`np_id`, `InChI`, `InChIKey`, and `chembl_id`) from `df_merged_chembl`.
*   The `InChI` column was successfully renamed to `standard_inchi`, and the `InChIKey` column was renamed to `standard_inchi_key` within `df_natural_products_chembl_info`.
*   The final DataFrame, `df_natural_products_chembl_info`, was saved as a CSV file named `natural_products_chembl_info.csv`, with the index excluded.

### Insights or Next Steps

*   The created `natural_products_chembl_info.csv` provides a consolidated and standardized dataset linking natural product identifiers to their InChI codes, InChIKeys, and ChEMBL IDs, ready for downstream analysis or database integration.
*   This dataset can now be used to query ChEMBL for detailed biological activity or target information related to these natural products, using the `chembl_id` column.


# Task
Install `chembl_downloader`, then load the NPASS natural products structure data from `/content/NPASS3.0_naturalproducts_structure.txt`, query ChEMBL for molecule dictionary information including InChI, InChIKey, and ChEMBL ID, and finally merge the NPASS structure data with the ChEMBL data to map NPASS compounds to ChEMBL IDs.

## Install chembl_downloader

### Subtask:
Install the missing `chembl_downloader` package.


**Reasoning**:
To install the missing `chembl_downloader` package, I will add the `!pip install chembl_downloader` command to the beginning of the specified code cell.



**Reasoning**:
The previous attempt to load `npass_structure_df` failed because of an incorrect file path. I need to correct the path from `/content/sample_data/NPASS3.0_naturalproducts_structure.txt` to `/content/NPASS3.0_naturalproducts_structure.txt`. The `chembl_downloader` is already installed, so I will remove the installation command.



## Summary:

### Data Analysis Key Findings

*   The `chembl_downloader` package was successfully installed.
*   NPASS Natural Products Structure data, containing 203,390 entries and 4 columns, was successfully loaded from `/content/NPASS3.0_naturalproducts_structure.txt`.
*   ChEMBL molecule dictionary information, including ChEMBL ID, InChI, and InChIKey, was queried, resulting in 2,854,815 entries and 3 columns.
*   The NPASS structure data was successfully merged with the ChEMBL data, creating a final DataFrame of 203,390 entries and 6 columns.
*   A total of 53,376 unique NPASS compounds were successfully mapped to ChEMBL IDs.

### Insights or Next Steps

*   The merging process successfully linked a significant portion of NPASS natural products to ChEMBL entries, enabling further analysis by leveraging the extensive bioactivity and chemical information available in ChEMBL.
*   For NPASS compounds not mapped to ChEMBL IDs, explore alternative methods for identification, such as substructure searching or cheminformatic tools, to maximize the number of integrated compounds.


# Task
Query ChEMBL for target information, selecting `chembl_id` and `uniprot_id` from the `target_dictionary` and `component_sequences` tables. Then, clean the `uniprot_id` column in `df_target` and merge it with the retrieved ChEMBL target data. Finally, select `target_id`, `target_name`, `target_type`, and `chembl_id`, display the first 5 rows, and save the result to `target_chembl_info.csv`.

## Query ChEMBL for Target Information

### Subtask:
Connect to the ChEMBL database and query for target `chembl_id` and `uniprot_id` by joining `target_dictionary` and `component_sequences` tables.


**Reasoning**:
To query the ChEMBL database for target information, I need to establish a connection, construct and execute a SQL query, and then display the results. I will use the `chembl_path` variable that was previously defined.



**Reasoning**:
The previous query failed because the `component_sequences` table does not contain the `uniprot_id` column, and the join condition was incorrect. The `uniprot_id` is available in the `target_components` table, which needs to be joined with `target_dictionary` on `tid` and `target_id` respectively to retrieve the `chembl_id` and `uniprot_id`.



**Reasoning**:
The previous queries failed because the `uniprot_id` column was not directly available in `target_components` or `component_sequences` tables as previously assumed. I need to join `target_dictionary` with `target_components` and then with `component_sequences` to correctly retrieve the UniProt accession, which is typically stored in the `accession` column of the `component_sequences` table, and alias it as `uniprot_id`.



**Reasoning**:
The previous query failed because the join condition between `target_dictionary` and `target_components` used an incorrect column name (`tc.target_id` instead of `tc.tid`). I will correct the join condition to `td.tid = tc.tid` to properly link the tables and retrieve the `chembl_id` and `uniprot_id`.



In [ ]:
import sqlite3

# Ensure chembl_path is defined (it should be from previous steps)
if 'chembl_path' not in locals():
    print("Warning: 'chembl_path' not found. Attempting to download ChEMBL again.")
    import chembl_downloader
    chembl_path = chembl_downloader.download_extract_sqlite(version="36")
    print(f"ChEMBL database path: {chembl_path}")

# 1. Establish a connection to the ChEMBL SQLite database
if chembl_path:
    conn = sqlite3.connect(chembl_path)

    # 2. Write a SQL query to get chembl_id and uniprot_id by joining target_dictionary, target_components, and component_sequences
    chembl_target_query = """
    SELECT
        td.chembl_id,
        cs.accession AS uniprot_id
    FROM
        target_dictionary td
    JOIN
        target_components tc ON td.tid = tc.tid
    JOIN
        component_sequences cs ON tc.component_id = cs.component_id
    WHERE
        cs.accession IS NOT NULL AND cs.accession != ''
    """

    # 3. Execute the SQL query and store results in a DataFrame
    chembl_targets_df = pd.read_sql_query(chembl_target_query, conn)

    # 4. Close the database connection
    conn.close()

    print("\nChEMBL target data queried successfully.")
    print(f"Shape: {chembl_targets_df.shape}")
    print("First 5 rows of ChEMBL target data:")
    display(chembl_targets_df.head())
else:
    print("Cannot query ChEMBL: chembl_path is not defined or invalid.")
    chembl_targets_df = pd.DataFrame()



ChEMBL target data queried successfully.
Shape: (16675, 2)
First 5 rows of ChEMBL target data:


,chembl_id,uniprot_id
0,CHEMBL2074,O43451
1,CHEMBL1971,O60706
2,CHEMBL1827,O76074
3,CHEMBL1859,O95180
4,CHEMBL1884,O96760


In [ ]:
print(df_target['uniprot_id'].value_counts(dropna=False).head())
print(df_target['uniprot_id'].unique()[:10])

uniprot_id
n.a.      3708
n.a        216
Q16445       5
P63138       4
P35372       4
Name: count, dtype: int64
['n.a.' 'O75496' 'O94925' 'Q99816' 'Q7Z2H8' 'P24046' 'P43220' 'P35375'
 'Q62053' 'P30557']


In [ ]:
import numpy as np

# Strip whitespace and replace all forms of 'n.a.' and 'nan' string with np.nan
df_target['uniprot_id'] = df_target['uniprot_id'].astype(str).str.strip()
df_target['uniprot_id'] = df_target['uniprot_id'].replace(['n.a.', 'n.a', 'nan'], np.nan)

print(df_target['uniprot_id'].value_counts(dropna=False).head())
print(df_target['uniprot_id'].unique()[:10])

uniprot_id
NaN       3924
Q16445       5
P35372       4
P28647       4
P63138       4
Name: count, dtype: int64
[nan 'O75496' 'O94925' 'Q99816' 'Q7Z2H8' 'P24046' 'P43220' 'P35375'
 'Q62053' 'P30557']


In [ ]:
df_target_merged_chembl = pd.merge(
    df_target,
    chembl_targets_df,
    left_on='uniprot_id',
    right_on='uniprot_id',
    how='left'
)

print("Shape of df_target_merged_chembl:", df_target_merged_chembl.shape)
display(df_target_merged_chembl.head())

Shape of df_target_merged_chembl: (12700, 7)


,target_id,target_type,target_name,target_organism_tax_id,target_organism,uniprot_id,chembl_id
0,NPT1031,Cell line,Ca9-22,9606,Homo sapiens,NaN,NaN
1,NPT1034,Cell line,Lu1,9606,Homo sapiens,NaN,NaN
2,NPT1045,Cell line,U2OS,9606,Homo sapiens,NaN,NaN
3,NPT1059,Cell line,HMEC,9606,Homo sapiens,NaN,NaN
4,NPT1076,Cell line,B-cells,n.a.,n.a.,NaN,NaN


In [ ]:
df_target_chembl_info = df_target_merged_chembl[[
    'target_id',
    'target_name',
    'target_type',
    'chembl_id'
]].copy()

print("Shape of df_target_chembl_info:", df_target_chembl_info.shape)
display(df_target_chembl_info.head())

df_target_chembl_info.to_csv('target_chembl_info.csv', index=False)
print('File target_chembl_info.csv has been saved.')

Shape of df_target_chembl_info: (12700, 4)


,target_id,target_name,target_type,chembl_id
0,NPT1031,Ca9-22,Cell line,NaN
1,NPT1034,Lu1,Cell line,NaN
2,NPT1045,U2OS,Cell line,NaN
3,NPT1059,HMEC,Cell line,NaN
4,NPT1076,B-cells,Cell line,NaN


File target_chembl_info.csv has been saved.


In [ ]:
if 'chembl_targets_df' in locals() and not chembl_targets_df.empty:
    print('DataFrame chembl_targets_df exists and is not empty.')
    print('Shape of chembl_targets_df:', chembl_targets_df.shape)
    print('Columns of chembl_targets_df:', chembl_targets_df.columns.tolist())
    print('First 5 rows of chembl_targets_df:')
    display(chembl_targets_df.head())

    # Verify specific columns
    if 'chembl_id' in chembl_targets_df.columns and 'uniprot_id' in chembl_targets_df.columns:
        print("Both 'chembl_id' and 'uniprot_id' columns are present.")
    else:
        print("Warning: Expected columns 'chembl_id' or 'uniprot_id' are missing.")
else:
    print('DataFrame chembl_targets_df does not exist or is empty.')

DataFrame chembl_targets_df exists and is not empty.
Shape of chembl_targets_df: (16675, 2)
Columns of chembl_targets_df: ['chembl_id', 'uniprot_id']
First 5 rows of chembl_targets_df:


,chembl_id,uniprot_id
0,CHEMBL2074,O43451
1,CHEMBL1971,O60706
2,CHEMBL1827,O76074
3,CHEMBL1859,O95180
4,CHEMBL1884,O96760


Both 'chembl_id' and 'uniprot_id' columns are present.


In [ ]:
print('First 5 rows of df_target_chembl_info:')
display(df_target_chembl_info.head())

print('\nShape of df_target_chembl_info:', df_target_chembl_info.shape)

print('\nColumns of df_target_chembl_info:', df_target_chembl_info.columns.tolist())

First 5 rows of df_target_chembl_info:


,target_id,target_name,target_type,chembl_id
0,NPT1031,Ca9-22,Cell line,NaN
1,NPT1034,Lu1,Cell line,NaN
2,NPT1045,U2OS,Cell line,NaN
3,NPT1059,HMEC,Cell line,NaN
4,NPT1076,B-cells,Cell line,NaN



Shape of df_target_chembl_info: (12700, 4)

Columns of df_target_chembl_info: ['target_id', 'target_name', 'target_type', 'chembl_id']
